### ISSUE TITLE
Rule no-unused-modules does not count re-export all declarations

### ISSUE THREAD

**Issue by DMartens** [benmosher/eslint-plugin-import#1945]
The rule `no-unused-modules` reports files which only have re-export all declarations for not having exports, eg:
```javascript
export * from './foo.js';
export * from './bar.js';
```

This seems to be intentional as the re-exports are [deleted from the count before the check](https://github.com/benmosher/eslint-plugin-import/blob/957092a8ecfd479c524bb0fa5d37ba9f5921bd5e/src/rules/no-unused-modules.js#L473).
I did not find any information on the reasoning for this behavior.  

My proposed solution is to add an option flag, eg. `allowReExportOnly`, to allow this.
The rule would still report if not a single re-export has an export.
I am willing to create a PR for this functionality if this change is approved.

**Comment by ljharb** [comment]
The reasoning I cam surmise is that there's no purpose in having this file - the consumers of it can (and should) be directly importing from `foo.js` or `bar.js` (I don't see any discussion about it in #1142 tho).

A PR for an option is reasonable, but when enabled, it should actually check that somebody is importing from that file - iow, if the file in your OP is `index.js`, and nobody is importing from `index.js`, then that file should be reported as unused.

Similarly, if that option is enabled, and only one of the `foo.js` exports (let's say "baz") is imported from `index.js`, and none of the `bar.js` exports, I would expect the rule to error until `index.js` has been reduced to `export { baz } from './foo.js'`.

**Comment by ljharb** [comment]
(Do note that barrel exports are an antipattern that drastically increases app footprint and that treeshaking is incapable of wholly avoiding - use deep imports instead)

**Comment by d8corp** [comment]
@ljharb 

71% drop is impressive for RN + large codebase. For single-export local barrels in web (ESM+Vite/esbuild), benchmarks show 1-5% overhead max.

Editor autocomplete helps discovery, not DRY export duplication that your rule forces.

All barrel "cons" are purely bundler analysis limits—if bundlers properly cached re-export resolution (as they do for other static analysis), only first-time cost remains. Modern tooling should evolve past punishing simple patterns.

Simple barrels = libraries' public API pattern. Total ban breaks real-world DX unnecessarily.
